pkg/cloud/services/networking/router.go

```diff
@@ -185,41 +186,36 @@ func setRouterExternalIPs(client *gophercloud.ServiceClient, openStackCluster *i
 	return nil
 }
 
-func (s *Service) DeleteRouter(openStackCluster *infrav1.OpenStackCluster, network *infrav1.Network) error {
-	if network.Router == nil || network.Router.ID == "" {
-		s.logger.V(4).Info("No need to delete router since no router exists.")
-		return nil
-	}
-	exists, err := s.existsRouter(network.Router.ID)
+func (s *Service) DeleteRouter(openStackCluster *infrav1.OpenStackCluster, clusterName string) error {
+	router, err := s.getRouter(clusterName)
 	if err != nil {
 		return err
 	}
-	if !exists {
-		s.logger.Info("Skipping router deletion because router doesn't exist", "router", network.Router.ID)
+
+	if router == nil || router.router.ID == "" {
 		return nil
 	}
-	if network.Subnet == nil || network.Subnet.ID == "" {
-		s.logger.V(4).Info("Skipping removing router interface since no subnet exists.")
-	} else {
-		_, err = routers.RemoveInterface(s.client, network.Router.ID, routers.RemoveInterfaceOpts{
-			SubnetID: network.Subnet.ID,
+
+	if router.subnet.ID != "" {
+		_, err = routers.RemoveInterface(s.client, router.router.ID, routers.RemoveInterfaceOpts{
+			SubnetID: router.subnet.ID,
 		}).Extract()
 		if err != nil {
-			if errors.IsNotFound(err) {
-				s.logger.V(4).Info("Router Interface already removed, no actions", "id", network.Router.ID)
-				return nil
+			if capoerrors.IsNotFound(err) {
+				s.logger.V(4).Info("Router Interface already removed, no actions", "id", router.router.ID)
+				// nothing to do
 			}
 			return fmt.Errorf("unable to remove router interface: %v", err)
```

Comment:
I think we have to do it this way:
```go
		if err != nil {
			if !capoerrors.IsNotFound(err) {
				return fmt.Errorf("unable to remove router interface: %v", err)
			}
			s.logger.V(4).Info("Router Interface already removed, nothing to do", "id", router.router.ID)
		} else {
			s.logger.V(4).Info("Removed RouterInterface of Router", "id", router.router.ID)
		}
```
I think that's the best way, with the current implementation we would return an error after `// nothing to do`